examples/experimental/imagenet_nas_arch_pytorch/model_def.py

```diff
@@ -0,0 +1,208 @@
+        )
+
+        valid_queue = DataLoader(
+            valid_data,
+            batch_size=self.context.get_per_slot_batch_size(),
+            shuffle=False,
+            pin_memory=True,
+            num_workers=2,
+        )
+        return valid_queue
+
+    def build_model(self) -> nn.Module:
+        genotype = Genotype(
+            normal=[
+                ('skip_connect', 1),
+                ('skip_connect', 0),
+                ('sep_conv_3x3', 2),
+                ('sep_conv_3x3', 1),
+                ('sep_conv_5x5', 2),
+                ('sep_conv_3x3', 0),
+                ('sep_conv_5x5', 3),
+                ('sep_conv_5x5', 2)
+            ], normal_concat=range(2,6),
+            reduce=[
+                ('max_pool_3x3', 1),
+                ('sep_conv_3x3', 0),
+                ('sep_conv_5x5', 1),
+                ('dil_conv_5x5', 2),
+                ('sep_conv_3x3', 1),
+                ('sep_conv_3x3', 3),
+                ('sep_conv_5x5', 1),
+                ('max_pool_3x3', 2)
+            ], reduce_concat=range(2,6)
+        )
+        activation_function = activation_map[self.context.get_hparam("activation")]
+
+        model = NetworkImageNet(
+            genotype,
+            activation_function,
+            self.context.get_hparam("init_channels"),
+            1000, # number of classes
+            self.context.get_hparam("layers"),
+            auxiliary=self.context.get_hparam("auxiliary"),
+            do_SE=self.context.get_hparam("do_SE")
+        )
+
+        # If loading backbone weights, do not call reset_parameters() or
+        # call before loading the backbone weights.
+        reset_parameters(model)
+        return model
+
+    def optimizer(self, model: nn.Module) -> torch.optim.Optimizer:  # type: ignore
+        self._optimizer = torch.optim.SGD(
+            model.parameters(),
+            lr=self.context.get_hparam("learning_rate"),
+            momentum=self.context.get_hparam("momentum"),
+            weight_decay=self.context.get_hparam("weight_decay")
+        )
+        return self._optimizer
+
+    def create_lr_scheduler(self, optimizer):
+        self.scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
+            optimizer, 300
+        )
+        step_mode = LRScheduler.StepMode.MANUAL_STEP
+        return LRScheduler(self.scheduler, step_mode=step_mode)
+
+    def train_batch(
+        self, batch: Any, model: nn.Module, epoch_idx: int, batch_idx: int
+    ) -> Dict[str, torch.Tensor]:
+
+        if batch_idx==0 or self.last_epoch_idx < epoch_idx:
+            self.scheduler.step()
+            current_lr = self.scheduler.get_last_lr()[0]
+
+            if epoch_idx < 5:
+                lr = self.context.get_hparam("learning_rate")
+                for param_group in self._optimizer.param_groups:
+                    param_group["lr"] = lr * (epoch_idx + 1) / 5.0
+                print("Warming-up Epoch: {}, LR: {}".format(epoch_idx, lr * (epoch_idx + 1) / 5.0))
```

Comment:
We may want to consider putting this into a LRschduler class so we can just pass in a LRschduler and have determined handle the scheduler for us. 

Did this produce the correct numbers in dtrain setting? I think we create a copy of the optimizer in horovod which means this would not be the same optimizer used for training. 
`            self.optimizer = hvd.DistributedOptimizer(
                self.optimizer,
                ....)`

The above LR scheduler solution should also solve this problem for us.

Reply:
This does not create a problem for dtrain because every instance of optimizer calls this (as evidenced by the log printing the warmup lr 8 times).  It might be cleaner to implement in a custom scheduler but the original PC-DARTS codebase also coded the warmup lr like this so I think it's okay to keep as is.